### PR TITLE
xds: nack route configuration with regexes that don't compile

### PR DESCRIPTION
**Pull request by easwars**
Fixes: https://github.com/grpc/grpc-go/issues/4240
#fixit

### CODE REVIEW

File: xds/internal/client/xds.go [hunk 1]
```diff
@@ -452,7 +458,12 @@ func routesProtoToSlice(routes []*v3routepb.Route, logger *grpclog.PrefixLogger,
 			case *v3routepb.HeaderMatcher_ExactMatch:
 				header.ExactMatch = &ht.ExactMatch
 			case *v3routepb.HeaderMatcher_SafeRegexMatch:
-				header.RegexMatch = &ht.SafeRegexMatch.Regex
+				regex := ht.SafeRegexMatch.GetRegex()
```
Comment (menghanl) [inline xds/internal/client/xds.go:461]: I was thinking we can be a bit more aggressive here.

Options:
1. (probably not a good idea): we turn the match fields into a [`matcher`](https://github.com/grpc/grpc-go/blob/d426aa5f2e5e809639b45d9619416ce22e56319a/xds/internal/resolver/matcher.go#L33), and put that in the RDS update struct, so the resolver can use directly.
   - this is bad because the log probably won't be as easy to read, and also it may be hard to compare RDS updates in the tests
1. (better and easier than 1?): change the type of `header.RegexMatch` to `regexp.Regexp`
   - this saves the `regexp.Compile()` in resolver (and the error handling)
   - this has the similar problem for tests. But I think we can just compare the result of `String()`

**Review by menghanl — CHANGES_REQUESTED** [review]

File: xds/internal/client/xds.go (hunk 1 above)
Comment (easwars) [reply xds/internal/client/xds.go:461]: Done.

Comment (menghanl) [reply xds/internal/client/xds.go:461]: Hmm, I'm surprised that we don't need to fix tests for this (the newer commit). Maybe we don't have Regex success tests, and the two Regex failure tests aren't affected. Can you double check?

File: xds/internal/client/client.go [hunk 2]
```diff
@@ -304,14 +307,14 @@ type WeightedCluster struct {
...
-	ExactMatch   *string     `json:"exactMatch,omitempty"`
-	RegexMatch   *string     `json:"regexMatch,omitempty"`
-	PrefixMatch  *string     `json:"prefixMatch,omitempty"`
-	SuffixMatch  *string     `json:"suffixMatch,omitempty"`
-	RangeMatch   *Int64Range `json:"rangeMatch,omitempty"`
-	PresentMatch *bool       `json:"presentMatch,omitempty"`
+	Name         string         `json:"name"`
+	InvertMatch  *bool          `json:"invertMatch,omitempty"`
+	ExactMatch   *string        `json:"exactMatch,omitempty"`
+	RegexMatch   *regexp.Regexp `json:"regexMatch,omitempty"`
```
Comment (menghanl) [inline xds/internal/client/client.go:313]: I don't know what JSON this field will be marshalled to.
But I don't think we actually care now.

We needed the json because the matchers were specified in balancer config, and passed to the routing balancer. Now that this is no longer marshaled to JSON, I think we can delete all the `json` tags.

We can do that in another PR.

**Review by menghanl — APPROVED** [review]

File: xds/internal/client/xds.go (hunk 1 above)
Comment (easwars) [reply xds/internal/client/xds.go:461]: Added a test for the regex success case.

**Review by menghanl — APPROVED** [review]
